File: Day #110/solution.py
# Day 110: Root-to-leaf paths via DFS backtracking. O(n) nodes, O(h) stack.
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def root_to_leaf(root):
    res = []

    def dfs(n, path):
        if not n:
            return
        path.append(n.val)
        if not n.left and not n.right:
            res.append(list(path))
        else:
            dfs(n.left, path)
            dfs(n.right, path)
        path.pop()

    dfs(root, [])
    return res


if __name__ == "__main__":
    root = Node(1, Node(2), Node(3, Node(4), Node(5)))
    print(root_to_leaf(root))

File: Day #110/solution.js
// Day 110: Root-to-leaf paths via DFS backtracking. O(n) nodes, O(h) stack.
class Node {
  constructor(val, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

function rootToLeaf(root) {
  const res = [];
  const dfs = (n, path) => {
    if (!n) return;
    path.push(n.val);
    if (!n.left && !n.right) res.push([...path]);
    else {
      dfs(n.left, path);
      dfs(n.right, path);
    }
    path.pop();
  };
  dfs(root, []);
  return res;
}

const root = new Node(1, new Node(2), new Node(3, new Node(4), new Node(5)));
console.log(JSON.stringify(rootToLeaf(root)).replace(/,/g, ", "));
